lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php

```diff
@@ -228,6 +247,50 @@ public function execute($params = null)
         return true;
     }
 
+    /**
+     * Handle $this->_longData after regular query parameters have been bound
+     *
+     * @throws MysqliException
+     */
+    private function processLongData()
+    {
+        foreach ($this->longData as $paramNr => $value) {
+            if (!is_resource($value)) {
```

Comment:
From https://github.com/doctrine/dbal/pull/3217#discussion_r216139584:

> Might lose the advantage of sending the parameter in a network package by itself, though.

This did make sense since I thought one could send a string which exceeds `max_allowed_packet` using `send_long_data()`. Given it's not the case, what's the advantage of sending the parameter in a network package by itself? Isn't it better to just bind the string as is and let the driver decide how to send it?

Reply:
Seems to make a difference regarding memory consumption, but that must happen down inside the PHP core/mysqli driver.

Please confirm by trying the following:

Patch the `BlobTest` like so:

```
diff --git a/tests/Doctrine/Tests/DBAL/Functional/BlobTest.php b/tests/Doctrine/Tests/DBAL/Functional/BlobTest.php
index 2c41ed89e..bc90c679d 100644
--- a/tests/Doctrine/Tests/DBAL/Functional/BlobTest.php
+++ b/tests/Doctrine/Tests/DBAL/Functional/BlobTest.php
@@ -37,17 +37,20 @@ class BlobTest extends \Doctrine\Tests\DbalFunctionalTestCase

     public function testInsert()
     {
+        $lob = str_repeat('x', 8 * 1024 * 1024);
         $ret = $this->_conn->insert('blob_table', [
             'id'          => 1,
-            'clobfield'   => 'test',
-            'blobfield'   => 'test',
+            'clobfield'   => $lob,
+            'blobfield'   => $lob,
         ], [
             ParameterType::INTEGER,
-            ParameterType::LARGE_OBJECT,
-            ParameterType::LARGE_OBJECT,
+            ParameterType::STRING,
+            ParameterType::STRING,
         ]);

         self::assertEquals(1, $ret);
+
+        print memory_get_peak_usage() / (1024*1024);
     }

     public function testInsertProcessesStream()
```

This creates a 8 MB BLOB, inserts it as a string and effectively short-circuits the entire `send_long_data()` logic introduced in this PR. At the end, it prints the peak memory usage, which for me was around 44 MB.

Then, change back to `ParameterType::LARGE_OBJECT`. For me, memory usage drops to 28 MB.

As the difference is two times the size of the LOB, my assumption is that somewhere down inside PHP, the MySQL driver must copy together all the data to be sent to MySQL into some buffer. Maybe `send_long_data()` can prevent this by sending out the data immediately.

Bottom line: I'd like to keep it this way as it leaves it to the user to decide whether to try any advanced optimizations, or just use `ParameterType::STRING` and pass in large strings. YMMV.

Reply:
The real problem is storing an 8 MB string in memory, not the way we send it to the DB.

I don't think we should solve this problem at the cost of complicating the API. In fact, every DB driver will have to be able to bind a string and a resource as `ParameterType::LARGE_OBJECT` which adds unnecessary complexity.

If memory usage critical is important but the data is already in memory, it can be dumped into a temporary file which then can be `fopen()`ed and bound to the statement.

Reply:
I don't get your point 😕. But re-reading my previous comment, I found it rather confusing and misunderstandable. So, let me try to restate it and I hope we can get along.

If you have your data – for whatever reason –  as a string in memory and pass it as the `STRING` type, there's not much we can do about it.

If you use the `LARGE_OBJECT` type to indicate your data is something "heavy", DBAL should do its best to handle it the most efficient way. 

If you pass in a stream, `mysqli` can process it in chunks. To make this work is what this PR is about. (Other platforms still lack support as the skipped tests show. Maybe the particular drivers can find efficient ways of handling streams, maybe they need to fall back to reading the entire stream and treat it as `STRING`. But all that is out of scope here.)

If you pass in a string, `mysqli` has a better memory footprint when using `send_long_data` instead of `bind_param`. This optimization is a side-effect of this PR, but it is also an internal detail of the DBAL MySQLi driver the user need not worry about.

So, is your suggestion that `LARGE_OBJECT` should no longer accept strings but only streams?

`BlobTest` suggests that passing strings as `LARGE_OBJECT` is supported for all platforms right now. It's stream support where we only have partial coverage (not DB2, not Oracle?)

Reply:
> So, is your suggestion that `LARGE_OBJECT` should no longer accept strings but only streams?

It should remain as is in order to not introduce breaking changes — a string bound as `LARGE_OBJECT` should be handled as a string, without using the `send_long_data()`. In the future, I'd like to deprecate the ability to bind strings as LOB. A string parameter should expect a string value, a LOB should expect a stream — just as simple as that.

> If you use the `LARGE_OBJECT` type to indicate your data is something "heavy", DBAL should do its best to handle it the most efficient way.

I don't agree here. DBAL doesn't have to optimize internal defects of a driver at the cost of increasing its own complexity. As a driver user, I'd indeed expect a large string bound to a statement to not increase the memory footprint significantly. If it's a bug, it should be fixed in the driver so that not only DBAL users could benefit from that.

> `BlobTest` suggests that passing strings as `LARGE_OBJECT` is supported for all platforms right now. It's stream support where we only have partial coverage (not DB2, not Oracle?)

This is just because the stream support is not yet implemented in other platforms, and it was relatively easy to implement handling strings. From the DB application standpoint, there's really no reason to represent LOBs as strings. They are just files stored in the DB.

Reply:
> But re-reading my previous comment, I found it rather confusing and misunderstandable. So, let me try to restate it and I hope we can get along.

I do completely get your point, I just don't agree.

Reply:
Changed. 

Once this is merged I can open another issue – maybe it helps someone else in the future or we might want to revisit this sometime.

Reply:
That sounds good to me. You may also want to file a ticket with `ext-mysqli` to get it potentially fixed in a proper place.